timemachine/md/smc.py

```diff
@@ -0,0 +1,179 @@
+# adapted from https://github.com/proteneer/timemachine/blob/7d6099b0f5b4a2d0b26c3edc7a91c18f7a526c00/md/experimental/smc.py
+
+from typing import Any, Callable, Dict, List, Tuple
+
+import numpy as np
+from scipy.special import logsumexp
+from typing_extensions import TypeAlias
+
+# type annotations
+Sample: TypeAlias = Any
+Samples: TypeAlias = List[Sample]  # e.g. List[CoordsVelBox]
+
+Lambda: TypeAlias = float
+LogWeight: TypeAlias = float
+Array: TypeAlias = np.ndarray
+IndexArray: TypeAlias = Array
+LogWeights: TypeAlias = Array
+
+BatchPropagator = Callable[[Samples, Lambda], Samples]
+BatchLogProb = Callable[[Samples, Lambda], LogWeights]
+
+Resampler = Callable[[LogWeights], Tuple[IndexArray, LogWeights]]
+ResultDict = Dict[str, Any]
+
+
+def sequential_monte_carlo(
+    samples: Samples,
+    lambdas: Array,
+    propagate: BatchPropagator,
+    log_prob: BatchLogProb,
+    resample: Resampler,
+) -> ResultDict:
+    """barebones implementation of Sequential Monte Carlo (SMC)
+
+    Parameters
+    ----------
+    samples: [N,] list
+    lambdas: [K,] array
+    propagate: function
+        [move(x, lam) for x in xs]
+        for example, move(x, lam) might mean "run 100 steps of all-atom MD targeting exp(-u(., lam)), initialized at x"
+    log_prob: function
+        [exp(-u(x, lam)) for x in xs]
+    resample: function
+        (optionally) perform resampling given an array of log weights
+
+    Returns
+    -------
+    trajs_dict
+        "sample_traj"
+            [K-1, N] list of snapshots
+        "incremental_log_weights_traj"
+            [K-1, N] array of incremental log weights
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+
+    References
+    ----------
+    * Arnaud Doucet's annotated bibliography of SMC
+        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
+    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
+        https://arxiv.org/abs/2007.11936
+    """
+    n = len(samples)
+    log_weights = np.zeros(n)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+        sample_traj.append(samples)
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+
+    # main loop
+    for (lam_initial, lam_target) in zip(lambdas[:-2], lambdas[1:-1]):
```

Comment:
This is a bit unintuitive to me. Why don't we target `lambdas[-1]`?

Reply:
Can you add a comment here, in case I forget our discussion?

Reply:
Can you add a comment here, in case I forget our discussion?

Reply:
Good call -- added comment and cross-references in https://github.com/proteneer/timemachine/pull/718/commits/272241d15cd37214d6cd9d662bdcd5492e067416